src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,291 @@
+
+                FileInfo responseFile = new FileInfo(Path.Combine(tempDirectory.FullName, Path.GetRandomFileName()));
+
+                try
+                {
+                    WriteResponseFile(responseFile.FullName);
+
+                    MSBuildLogger logger = new MSBuildLogger(Log);
+
+                    using (var semaphore = new SemaphoreSlim(initialCount: 0, maxCount: 1))
+                    using (var loggingQueue = new TaskLoggingQueue(Log))
+                    using (var process = new Process())
+                    {
+                        process.EnableRaisingEvents = true;
+                        process.StartInfo = new ProcessStartInfo
+                        {
+                            Arguments = $"\"{string.Join("\" \"", GetCommandLineArguments(responseFile))}\"",
+                            CreateNoWindow = true,
+                            FileName = GetProcessFileName(ProcessFileName),
+                            RedirectStandardOutput = true,
+                            UseShellExecute = false,
+                            WorkingDirectory = Environment.CurrentDirectory,
+                        };
+
+                        // Place the output in the queue which handles logging messages coming through on StdOut
+                        process.OutputDataReceived += (sender, args) => loggingQueue.Enqueue(args?.Data);
+
+                        process.Exited += (sender, args) => semaphore.Release();
+
+                        try
+                        {
+                            Log.LogMessage(MessageImportance.Low, "Running command: \"{0}\" {1}", process.StartInfo.FileName, process.StartInfo.Arguments);
+
+                            process.Start();
+
+                            process.BeginOutputReadLine();
+
+                            semaphore.Wait(_cancellationTokenSource.Token);
+
+                            if (!process.HasExited)
+                            {
+                                try
+                                {
+                                    process.Kill();
+                                }
+                                catch (InvalidOperationException)
+                                {
+                                    // The process may have exited, in this case ignore the exception
+                                }
+                            }
+                        }
+                        catch (Exception e) when (
+                            e is OperationCanceledException
+                            || (e is AggregateException aggregateException && aggregateException.InnerException is OperationCanceledException))
+                        {
+                        }
+                    }
+                }
+                finally
+                {
+                    FileUtility.Delete(responseFile.FullName);
```

Comment:
Thank you for deleting the file. Some people think files in the temp directory should be left to linger, and allow an operating system process to clean it up.

I'm wondering if for debugging reasons we might want to have an environment variable to not delete the file? That way we can tell customers to set the env var, run restore, then send us the file?

Reply:
I was thinking the same thing which is how to improve debuggability. In my experience, _most_ of the time there are enough logs emitted by the main MSBuild process and the NuGet.Build.Tasks.Console process (as long as everything works).  That said, there's no reason to not include a flag here to leave the file around just in case, so I'll add one.

Reply:
Added an escape hatch, `SET RESTORE_TASK_DELETE_ARGUMENT_FILE=false`